### PR TITLE
#149879334 Add before_insert hook to generate unique instance ids

### PR DESCRIPTION
**Pull request by LarryWachira**
#### What does this PR do?
Adds a `before_insert` event hook to generate unique string identifiers for the tables
#### How should this be manually tested?
At the root of the project directory run the following command:
        
    $ pytest -v

#### What are the relevant pivotal tracker stories?
#149879334

### CODE REVIEW

File: api/models.py [hunk 1]
```diff
@@ -152,3 +156,21 @@ class Path(db.Model, ModelOpsMixin):
...
+    '''
+    A function to generate unique identifiers on insert
+    '''
+    push_id = PushID()
+    target.id = push_id.next_id()
+
+
+# associate the listener function with models, to execute during the
+# "before_insert" hook
+event.listen(ContributionType, 'before_insert', fancy_id_generator)
```
Comment (Andela-ddiei) [inline api/models.py]: Make this DRY.

File: api/unique_id.py [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+    # Modeled after base64 web-safe chars, but ordered by ASCII.
+    PUSH_CHARS = ('-0123456789'
+                  'ABCDEFGHIJKLMNOPQRSTUVWXYZ'
+                  '_abcdefghijklmnopqrstuvwxyz')
+
+    def __init__(self):
+
+        # Timestamp of last push, used to prevent local collisions if you
+        # pushtwice in one ms.
+        self.lastPushTime = 0
```
Comment (Andela-ddiei) [inline api/unique_id.py]: Make all variable names pythony

File: api/unique_id.py [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+        now = int(time() * 1000)
+        duplicateTime = (now == self.lastPushTime)
+        self.lastPushTime = now
+        timeStampChars = numpy.empty(8, dtype=str)
+
+        for i in range(7, -1, -1):
+            timeStampChars[i] = self.PUSH_CHARS[now % 64]
+            now = int(now / 64)
+
+        if (now != 0):
```
Comment (Andela-ddiei) [inline api/unique_id.py]: `if now`

File: test/api/views/test_contributions.py [hunk 4]
```diff
@@ -147,20 +136,20 @@ def test_fetch_user_contributions(self):
...
+            )
 
         # NOTE: This query should be updated to fetch contribution types by
         # user_id
         all_contribution_types = len(set(ContributionType.query.all()))
 
         self.assertEqual(response_contribution_types, all_contribution_types)
 
-    def test_fetch_user_contributions(self):
+    def test_fetch_user2_contributions(self):
```
Comment (Andela-ddiei) [inline test/api/views/test_contributions.py]: Change test name.

File: test/api/views/test_contributions.py [hunk 5]
```diff
@@ -445,24 +432,27 @@ def test_empty_request_data(self):
...
-            id=contrib_id,
             title="AI",
             user_id="-KJKcv0D",
-            contribution_type_id="-KSDQcv0D"
+            contribution_type=contrib_type
         )
-        contrib.contribution_type = self.resource
-        contrib.save()
-        old_contrib_status = Contribution.query.filter_by(id=contrib_id).first().status
+        db.session.add(contrib)
```
Comment (Andela-ddiei) [inline test/api/views/test_contributions.py]: contrib.save()

**Comment by Andela-ddiei** [comment]
LGTM.

**Comment by andela-oowonikoko** [comment]
LGTM

**Pull request merged**